builtins: Implement ST_LineFromEncodedPolyline function

mknycha opened the pull request:
This function takes the encoded polyline and creates linestring out of it.
Works the same as function in PostGIS.
Release note (sql change): Implement geo builtin ST_LineFromEncodedPolyline.

Resolves #48813

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55429)
<!-- Reviewable:end -->

otan commented on pkg/geo/encode.go:
```diff
@@ -311,3 +312,57 @@ func StringToByteOrder(s string) binary.ByteOrder {
 }
+
+// LineFromEncodedPolyline takes the encoded polyline ASCII and precision, decodes the points and returns them as a geometry
+func LineFromEncodedPolyline(encodedPolyline string, precision int) (Geometry, error) {
```
Two things:
* Can you move the polyline logic (decodePolylinePoints, decodePointValue) to polyline.go.
* Can you rename this to ParseEncodedPolyline, and move this to parse.go?

mknycha replied:
Ok, done

otan commented on pkg/geo/encode.go:
```diff
@@ -311,3 +312,57 @@ func StringToByteOrder(s string) binary.ByteOrder {
+	return g, nil
+}
+
+func decodePolylinePoints(encoded string, precision int) []float64 {
```
Can you add a comment to the function which describes how the algorithm works?

mknycha replied:
Done

mknycha replied:
But wait, the link to the algorithm is already added in the functions under `geo_builtins`. Do you want me to add it here anyway?

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -4827,6 +4827,56 @@ The swap_ordinate_string parameter is a 2-character string naming the ordinates
+		},
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{"encoded polyline", types.String},
```
nit: encoded_polyline.

mknycha replied:
Done

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -4827,6 +4827,56 @@ The swap_ordinate_string parameter is a 2-character string naming the ordinates
+	"st_linefromencodedpolyline": makeBuiltin(defProps(),
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{"encoded polyline", types.String},
```
nit: encoded_polyline.

mknycha replied:
Done

otan commented on pkg/geo/encode_test.go:
```diff
@@ -320,3 +320,101 @@ func TestSpatialObjectToGeoHash(t *testing.T) {
+			"CACC",
+			0,
+			"SRID=4326;LINESTRING(1 2, 3 4)",
+		},
```
for completelyness, can you add an empty string?

mknycha replied:
Good catch, added

otan reviewed (commented):
looking good! just some requested moves

otan reviewed (commented):
i think this needs a rebase, otherwise good to go!@

mknycha committed:
builtins: Implement ST_LineFromEncodedPolyline function

This function takes the encoded polyline and creates linestring out of it.
Works the same as function in PostGIS.
Release note (sql change): Implement geo builtin ST_LineFromEncodedPolyline.

mknycha commented:
It looks like linter is telling me `Ensure generated code is up-to-date`, but it is up to date. I am also not able to replicate this problem locally with `make lint`. What can I do about it?

mknycha commented on pkg/geo/polyline.go:
```diff
@@ -0,0 +1,56 @@
+import "math"
+
+// Reverse the polyline algorithm: https://developers.google.com/maps/documentation/utilities/polylinealgorithm
+func decodePolylinePoints(encoded string, precision int) []float64 {
```
Corrected
